src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -126,45 +170,50 @@ public final boolean hasMoreElements() {
+            enumCtx.setRes(newRes);
+            if (enumCtx.res == null) {
+                limit = posn = 0;
+                return;
+            }
 
-        res = homeCtx.getSearchReply(enumClnt, res);
-        if (res == null) {
-            limit = posn = 0;
-            return;
-        }
-
-        entries = res.entries;
-        limit = (entries == null) ? 0 : entries.size(); // handle empty set
-        posn = 0; // reset
+            entries = enumCtx.res.entries;
+            limit = (entries == null) ? 0 : entries.size(); // handle empty set
+            posn = 0; // reset
 
-        // minimize the number of calls to processReturnCode()
-        // (expensive when batchSize is small and there are many results)
-        if ((res.status != LdapClient.LDAP_SUCCESS) ||
-            ((res.status == LdapClient.LDAP_SUCCESS) &&
-                (res.referrals != null))) {
+            // minimize the number of calls to processReturnCode()
+            // (expensive when batchSize is small and there are many results)
+            if ((enumCtx.res.status != LdapClient.LDAP_SUCCESS) ||
+                ((enumCtx.res.status == LdapClient.LDAP_SUCCESS) &&
+                    (enumCtx.res.referrals != null))) {
 
-            try {
-                // convert referrals into a chain of LdapReferralException
-                homeCtx.processReturnCode(res, listArg);
+                try {
+                    // convert referrals into a chain of LdapReferralException
+                    homeCtx().processReturnCode(enumCtx.res, listArg);
 
-            } catch (LimitExceededException | PartialResultException e) {
-                setNamingException(e);
+                } catch (LimitExceededException | PartialResultException e) {
+                    setNamingException(e);
 
+                }
             }
-        }
 
-        // merge any newly received referrals with any current referrals
-        if (res.refEx != null) {
-            if (refEx == null) {
-                refEx = res.refEx;
-            } else {
-                refEx = refEx.appendUnprocessedReferrals(res.refEx);
+            // merge any newly received referrals with any current referrals
+            if (enumCtx.res.refEx != null) {
+                if (refEx == null) {
+                    refEx = enumCtx.res.refEx;
+                } else {
+                    refEx = refEx.appendUnprocessedReferrals(enumCtx.res.refEx);
+                }
+                enumCtx.res.refEx = null; // reset
             }
-            res.refEx = null; // reset
-        }
 
-        if (res.resControls != null) {
-            homeCtx.respCtls = res.resControls;
+            if (enumCtx.res.resControls != null) {
+                homeCtx().respCtls = enumCtx.res.resControls;
+            }
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(this);
```

Comment:
I don't think there is any benefit to the `try{} finally {fence}`.
The reachabilityFence has no executable code. Its only purpose is to keep the reference in scope alive.

Reply:
> I don't think there is any benefit to the `try{} finally {fence}`. The reachabilityFence has no executable code. Its only purpose is to keep the reference in scope alive.

...reachability fence has no code, but puting it into finally block makes the compiler see it on all possible exit paths from the try block (any possible exceptions, returns, etc.) so the argument 'this' is protected from being GC-ed until every possible code in try block is finished.

Reply:
The memory fence is not needed though (as explained above).